crates/core/src/metadata_store/providers/objstore/optimistic_store.rs

```diff
@@ -0,0 +1,254 @@
+            return Ok(());
+        }
+
+        let maybe_global_latest_version = self
+            .version_repository
+            .get_latest_version()
+            .await
+            .map_err(|e| anyhow!(e))?;
+
+        let maybe_cached_version = self
+            .latest_store_cache
+            .load()
+            .as_ref()
+            .as_ref()
+            .map(|cache| cache.store.current_version());
+
+        let Some(global_latest_version) = maybe_global_latest_version else {
+            // nothing to do, currently there is no known (or visible) latest version.
+            // we will optimistically create an empty store, assuming that there is no
+            // previous metadata stored.
+            if maybe_cached_version.is_none() {
+                self.latest_store_cache
+                    .store(Arc::new(Some(CachedStore::default())));
+            }
+            return Ok(());
+        };
+
+        if maybe_cached_version.is_none() || maybe_cached_version.unwrap() < global_latest_version {
+            tracing::info!(
+                "New metadata version discovered {} (previously known was {:?})",
+                global_latest_version,
+                maybe_cached_version
+            );
+            let latest_store_buf = self
+                .version_repository
+                .get_version(global_latest_version)
+                .await
+                .map_err(|e| anyhow!(e))?;
+
+            let last_store =
+                ImmutableStore::deserialize(latest_store_buf).map_err(WriteError::Network)?;
+
+            let cache = CachedStore::new(last_store);
+            self.latest_store_cache.store(Arc::new(Some(cache)));
+            return Ok(());
+        }
+        Ok(())
+    }
+
+    async fn get_or_update_store(&self) -> anyhow::Result<Arc<Option<CachedStore>>> {
+        self.maybe_update_cached_store(false).await?;
+        let maybe_current_store = self.latest_store_cache.load_full();
+        Ok(maybe_current_store)
+    }
+
+    async fn compare_and_swap(
+        &self,
+        mut op: impl FnMut(&ImmutableStore) -> Result<ImmutableStore, WriteError>,
+    ) -> Result<(), WriteError> {
+        for _retry_count in 0..25 {
+            let current_store = self
+                .get_or_update_store()
+                .await
+                .map_err(|e| WriteError::Network(e.into()))?;
+            let res = match current_store.as_ref() {
+                Some(cached_store) => op(&cached_store.store),
+                None => unreachable!(),
+            };
+            let Ok(new_store) = res else {
+                // a precondition failure. the actual failure is reflected in the Err.
+                return res.map(|_| ());
+            };
+
+            let new_store_bytes = new_store.serialize().map_err(WriteError::Store)?;
+
+            match self
+                .version_repository
+                .try_create_version(new_store.current_version(), new_store_bytes)
```

Comment:
The correctness of the implementation relies on the fact that our version space won't have any wholes, right? I guess this makes cleaning older versions up a bit more tricky.